test/Microsoft.Extensions.DependencyInjection.Tests/ServiceProviderExtensionsTest.cs

```diff
@@ -219,5 +256,25 @@ public interface IBar { }
         public class Bar1 : IBar { }
 
         public class Bar2 : IBar { }
+
+        private class RequiredServiceAwareServiceProvider : IServiceProvider, IRequiredServiceAware
+        {
+            public const string ErrorMessageFormat = "Boooo!! Required service {0} was not found";
+
+            object IRequiredServiceAware.GetRequiredService(Type serviceType)
+            {
+                if (serviceType == typeof(IBar))
+                {
+                    return new Bar1();
+                }
+
+                throw new InvalidOperationException(string.Format(ErrorMessageFormat, serviceType.Name));
```

Comment:
Much easier to just return a custom exception type. `throw new RankException` or some thing like that and you don't have to verify the exception message.